lib/volk_rank_archs.c

```diff
@@ -27,21 +27,43 @@
 #include <volk/volk_prefs.h>
 #include <volk_rank_archs.h>
 
-int volk_get_index(const char* impl_names[], // list of implementations by name
-                   const size_t n_impls,     // number of implementations available
-                   const char* impl_name     // the implementation name to find
-)
+// This function is supposed to be private to this compilation unit.
+int volk_search_index(const char* impl_names[],
+                      const size_t n_impls,
+                      const char* impl_name)
 {
     unsigned int i;
     for (i = 0; i < n_impls; i++) {
-        if (!strncmp(impl_names[i], impl_name, 20)) {
+        if (!strncmp(impl_names[i], impl_name, 42)) {
             return i;
         }
     }
-    // TODO return -1;
-    // something terrible should happen here
-    fprintf(stderr, "Volk warning: no arch found, returning generic impl\n");
-    return volk_get_index(impl_names, n_impls, "generic"); // but we'll fake it for now
+    return -1; // Indicate we couldn't find anything!
+}
+
+int volk_get_index(const char* impl_names[], // list of implementations by name
+                   const size_t n_impls,     // number of implementations available
+                   const char* impl_name     // the implementation name to find
+)
+{
+    /*
+     * We follow a 3 step process.
+     * 1. Search for the requested impl. Return index if found.
+     * 2. Search for the generic impl. Return as fail-safe.
+     * 3. Return -1 to indicate an error. Caller needs to handle this case.
+     */
+    int idx = volk_search_index(impl_names, n_impls, impl_name);
+    if (idx >= 0) {
+        return idx;
+    }
+
+    idx = volk_search_index(impl_names, n_impls, "generic");
+    if (idx >= 0) {
+        fprintf(stderr, "Volk warning: no arch found, returning generic impl\n");
+    } else {
+        fprintf(stderr, "Volk ERROR: no arch found. generic impl missing!\n");
+    }
+    return idx;
```

Comment:
we're just bubbling up the -1 here in case of error, but we need to handle that. Otherwise, we'll jump into some random memory location when the consumer program actually calls the kernel. That's worse than an infinite loop.

So, this is half the fix so far.

Reply:
at the very least, `__init_${kern.name}` needs to check for -1 and `abort` in that case. Failing early instead of wandering into random memory, where we become a stability or security issue, is necessary. Otherwise, we're just making bad code worse!

Reply:
also note that `__init_..` casts the `int` that `volk_rank_archs` returns into a `size_t`, which is unsigned.